#Gene region: DropSeq.eqtl::ploteQTLRegion
#(geneName, start, end, cisEQTLs, geneLoc, ratioSNPsToGenes=0.6, indexSNP=NULL, mafColumn=NULL)

#Individual SNP gene x expression plot: DropSeq.eqtl::plotGenotypesVsExpressionSimple
#(geneName, indexSNP, snpLoc, expressionMatrix, geneotypes, strTitlePrefix="", qvalue=NULL, ylim=NULL)

# library (DropSeq.eqtl)
# library (data.table)
# library (LDlinkR)
#Find all genes that the gwas SNP was tested in.
#for each gene, plot the region
#highlight the gwas SNP.
#for each gene plot the snp x gene expression for that gene and the INDEX SNP AND the GWAS SNP.

# base.dir="/downloads/d21STORMI"
# base.name="d21STORMI.maf_0.05_cisDist_250kb"
# gwasSNPs=c("chr15:78576056:C:T","chr15:78596058:G:A", "chr15:78556692:T:C")

# base.dir="/downloads/d21STORMI"
# base.name="d21Ngn2.maf_0.05_cisDist_250kb"
# gwasSNPs=c("15:78868398:C:T","15:78888400:G:A", "15:78849034:T:C")
# gwasSNPs=c("15:78868398:C:T")

# outDir=base.dir


# getSNPLocationFileName<-function (base.dir, base.name) paste(base.dir, "/", base.name, ".variant_locations.txt", sep="");
# getSNPFileName<-function (base.dir, base.name) paste(base.dir, "/", base.name , ".genotype_matrix.txt", sep="");
# getExpressionFileName<-function (base.dir, base.name) paste(base.dir, "/", base.name, ".gene_expression.txt", sep="");
# getGeneLocationFileName<-function (base.dir, base.name) paste(base.dir, "/", base.name, ".gene_locations.txt", sep="");
# getIndexSNPsFileName<-function (base.dir, base.name) paste(base.dir, "/", base.name, ".eQTL_eigenMT_results.txt", sep="");
# getEmpiricPvalueFileName<-function (base.dir, base.name) paste(base.dir, "/", base.name, ".eQTL_results.txt", sep="");

runGwasHitDefaultFileNames<-function (gwasSNPs, base.dir, base.name, outDir) {
    snps_location_file_name = getSNPLocationFileName(base.dir, base.name)
    SNP_file_name= getSNPFileName(base.dir, base.name)

    # Gene expression file name
    expression_file_name = getExpressionFileName(base.dir, base.name)
    gene_location_file_name = getGeneLocationFileName(base.dir, base.name)
    index_snps_file=getIndexSNPsFileName(base.dir, base.name)
    empiric_pvalues=getEmpiricPvalueFileName(base.dir, base.name)
    sapply(gwasSNPs, runOneGwasHit, gene_location_file_name, snps_location_file_name, expression_file_name, SNP_file_name,
                  index_snps_file, empiric_pvalues, outDir)

}



#' Scan/plot eQTL results by index SNP
#'
#' For a given SNP, search for all genes that were tested by that SNP in the empiric pvalues input file.  For each gene, plot the region
#' empiric pvalues as a manhatten plot. Also plot the eQTL index SNP and GWAS SNP genotype by expression interaction plot.
#'
#' @param gwasSNP The formatted location name of the GWAS SNP.  This is the chromosome:position:ref:alt, for example
#' "chr15:78576056:C:T"
#' @param gene_location_file_name The location of each gene.  Same number of lines as the expression_file_name, encodes the gene name, chromosome, start, end.
#' @param snps_location_file_name The location of each SNP.  Same number of lines as the SNP_file_name, encodes the snp name, chromosome, position.
#' @param expression_file_name A matrix of expression data, 1 row per gene, 1 column per sample.
#' @param SNP_file_name A matrix of genotypes, 1 row per SNP, 1 column per sample.  Encoded as 0,1,2 copies of the alternate allele.
#' @param index_snps_file The eQTL permuted results.  This is typically the eigenMT index file generated by the eQTL discovery pipeline - ie: eQTL_eigenMT_results.txt.
#' This should use the search window size and MAF appropriate for your sample size.
#' @param empiric_pvalues The eQTL empiric pvalues file.  This is typically the empric pvalues file from the eQTL discovery pipeline - ie: .eQTL_results.txt.  This
#' typically uses a more permissive search window size and MAF than the index_snps_file input.
#' @param outDir The output directory to generate the PDF report.  The report file is named for both the query SNP name as well as the basename of the input data.
#' @export
runOneGwasHit<-function (gwasSNP, gene_location_file_name, snps_location_file_name, expression_file_name, SNP_file_name,
                         index_snps_file, empiric_pvalues, outDir) {
    #Load up data:
    geneLoc=DropSeq.utilities::fastRead(gene_location_file_name, comment_regexp = "#")
    snpLoc=DropSeq.utilities::fastRead(snps_location_file_name, comment_regexp = "#")
    expressionMatrix=DropSeq.utilities::fastRead(expression_file_name, comment_regexp = "#")
    genotypes=DropSeq.utilities::fastRead(SNP_file_name, comment_regexp = "#")
    indexSNPs=DropSeq.utilities::fastRead(index_snps_file, comment_regexp = "#")
    allSNPs=DropSeq.utilities::fastRead(empiric_pvalues, comment_regexp = "#")

    geneList=unique (allSNPs[allSNPs$SNP==gwasSNP,]$gene)

    if (length(geneList)==0)
        stop (paste("No genes found overlapping SNP.  Maybe SNP wasn't tested by eQTL analysis?", gwasSNP))

    if (!is.null(outDir)) {
        baseName=sub(".gene_locations.txt", "", basename(gene_location_file_name))
        outFile=paste(outDir, "/", baseName, ".", gsub(":", "_", gwasSNP), ".pdf", sep="")
        pdf(outFile, width=10, height=7)
    }

    for (i in 1:length(geneList)) {
        cat (geneList[i],"\n")
        plotRegion(geneList[i], gwasSNP, allSNPs, geneLoc)
        plotSNPByGeneInteractions(geneList[i], gwasSNP, allSNPs, indexSNPs, snpLoc, expressionMatrix, genotypes)
    }
    if (!is.null(outDir)) dev.off()


}

getLDToEQTLHits<-function (geneList, gwasSNP, indexSNPs) {
    #this needs rsIDs!

    ii=indexSNPs[match(geneList, indexSNPs$gene),]

}

#geneName=geneList[1]
#if there's no index SNP, pick the minimum pvalue.
plotSNPByGeneInteractions<-function (geneName, gwasSNP, allSNPs, indexSNPs, snpLoc, expressionMatrix, genotypes) {
    #get the best hit for the gene, and plot that plus the gwas hit.
    #debug(DropSeq.eqtl::plotGenotypesVsExpression)
    eQTLBestHit = indexSNPs[indexSNPs$gene==geneName,]$SNP
    if (length(eQTLBestHit)==0) {
        temp=allSNPs[allSNPs$gene==geneName,]
        eQTLBestHit=temp[which.min(temp$`p-value`),]$SNP
    }
    par (mar=c(4,4,4,4), mfcol=c(1,2))
    DropSeq.eqtl::plotGenotypesVsExpression(geneName, allSNPs, snpLoc, expressionMatrix, genotypes, strTitlePrefix=eQTLBestHit, indexSNP=eQTLBestHit, ylim=NULL)
    DropSeq.eqtl::plotGenotypesVsExpression(geneName, allSNPs, snpLoc, expressionMatrix, genotypes, strTitlePrefix=gwasSNP, indexSNP=gwasSNP, ylim=NULL)

    #(geneName, indexSNP, snpLoc, expressionMatrix, geneotypes, strTitlePrefix="", qvalue=NULL, ylim=NULL)


}

#geneName=geneList[1]
plotRegion<-function (geneName, gwasSNP, allSNPs, geneLoc) {
    posRange=range (allSNPs[allSNPs$gene==geneName,]$snp_pos)
    start=posRange[1]
    end=posRange[2]
    z=strsplit (gwasSNP, ":")[[1]]
    indexSNP=data.frame(chr=z[1], pos=z[2], a1=z[3], a2=z[4], stringsAsFactors = F)
    #(geneName, start, end, cisEQTLs, geneLoc, ratioSNPsToGenes=0.6, indexSNP=NULL, mafColumn=NULL)
    #undebug (DropSeq.eqtl::ploteQTLRegion)
    z=DropSeq.eqtl::ploteQTLRegion(geneName, start, end, allSNPs, geneLoc, indexSNP=indexSNP, mafColumn="MAF")
    return (NULL)
}



